### PR TITLE
added audit-logs command

### PR DESCRIPTION
**Pull request by kiran-chaudhary**
**Description of Change**

Added `audit-logs` command in cli.

closes  #1179

**Testing Procedure**

`code42 audit-logs search -b 2020-08-09 --usernames 'username1' --usernames 'username2'`

**PR Checklist**

Did you remember to do the below?

- [x] Add unit tests to verify this change
- [x] Add an entry to CHANGELOG.md describing this change
- [x]  Add docstrings for any new public parameters / methods / classes

### CODE REVIEW

File: src/code42cli/cmds/auditlogs.py
Comment (kiran-chaudhary) [inline src/code42cli/cmds/auditlogs.py:10]: since we set `multiple=True`, it allows us to specify more than one usernames in format....
`--usernames "u1" --usernames "u2" ` in such case should the argument name be singular or plural?? e.g `--usernames` or `--username`

File: src/code42cli/cmds/auditlogs.py
Comment (kiran-chaudhary) [inline src/code42cli/cmds/auditlogs.py:118]: Need to remove this case based on discussion in standup.

File: tests/cmds/test_auditlogs.py
Comment (kiran-chaudhary) [inline tests/cmds/test_auditlogs.py:87]: `info` method is called but still it says called 0 times. Looking into it.

File: src/code42cli/cmds/search/extraction.py
Comment (kiran-chaudhary) [inline src/code42cli/cmds/search/extraction.py:167]: Open to suggestions for an apt name here.

File: src/code42cli/cmds/auditlogs.py
Comment (kiran-chaudhary) [reply src/code42cli/cmds/auditlogs.py]: removed.

File: src/code42cli/cmds/auditlogs.py
Comment (alanag13) [reply src/code42cli/cmds/auditlogs.py:15]: I would stick with the same convention that we have on other commands. I believe everything else uses singular names, so we should be consistent and do that here, too.

File: src/code42cli/cmds/auditlogs.py
Comment (alanag13) [inline src/code42cli/cmds/auditlogs.py:46]: i think that begin / end time arguments exist in another place in the codebase, can we import them and share those definitions instead of copying them over here? Are the time formats for audit log the same as what is used for file events / alerts?

**Comment by alanag13** [comment]
@kiran-chaudhary the minimum version of py42 will need to be bumped to be whatever the version is once the audit log functionality is released there.

File: src/code42cli/cmds/auditlogs.py
Comment (kiran-chaudhary) [reply src/code42cli/cmds/auditlogs.py]: Refactored, there are subtle differences in usage so had skipped it in the first iteration. Thanks.

File: src/code42cli/cmds/auditlogs.py
Comment (kiran-chaudhary) [reply src/code42cli/cmds/auditlogs.py]: changed.

**Comment by kiran-chaudhary** [comment]
> @kiran-chaudhary the minimum version of py42 will need to be bumped to be whatever the version is once the audit log functionality is released there.

Thanks, yes I will merge after approval, until py42 is released.
> 
> one thing that I understand but is still pretty scary to me: this build does not use a version of py42 that has audit log in it, but all the tests still passed. I think that's a sign that we're relying on mocking a little too heavily.

If we add unit tests that require the necessary build dependencies available, it will block development.  I tested the same by pulling the required py42 code from the audit-logs branch. We shouldn't move it to testing until the version is bumped and integration test should be written to test such cases and dependencies.

**Comment by unparalleled-js** [comment]
Should we add a required begin date like we do for file events?

**Comment by unparalleled-js** [comment]
Should we tack on the `csv` format option?

**Comment by unparalleled-js** [comment]
I am wondering why there is no Table output format as the default, rather than just json objects?

**Comment by kiran-chaudhary** [comment]
> Should we tack on the `csv` format option?

`csv` and `cef` format are supported by the API itself, While implementing those in py42, we observed those options do not return paginated response, etc and hence had decided to skip them for now. I thought we would be following same approach over here.

**Comment by kiran-chaudhary** [comment]
> There are a lot of results sometimes, should we use `echo_via_pager()` again if the result count is more than 10?

Yes, I can add that.

**Comment by kiran-chaudhary** [comment]
> I am wondering why there is no Table output format as the default, rather than just json objects?

Yes, I am open to supporting them as we have done for other commands if we do not want to use the API supported format options.

**Comment by alanag13** [comment]
I agree with adding all the items that @unparalleled-js called out.
Let's plan to cut a py42 release soon so that we can get the appropriate version set on this.

**Comment by unparalleled-js** [comment]
I can get a strange error when I use `-e`  on its own without `-b`.  This action does not appear to be allowed by the server.  We should not allow this in the CLI and echo some error text when the user tries.

**Comment by unparalleled-js** [comment]
When I try to use a format such as `2020-02-02` (not time), I can a strange error and we send `null` in the reqest... We should tack on zero seconds in this case so that this easy format is allowed.

**Comment by unparalleled-js** [comment]
It appears that the `-b` and `-e` flags do not work. It always sends null in the request when I try this.

**Comment by kiran-chaudhary** [comment]
> There are a lot of results sometimes, should we use `echo_via_pager()` again if the result count is more than 10?

Added.

**Comment by kiran-chaudhary** [comment]
> Should we tack on the `csv` format option?

Added.

**Comment by kiran-chaudhary** [comment]
> I am wondering why there is no Table output format as the default, rather than just json objects?

Added, have not defined headers, let me know the fields if needed, I think the "request json" can be done away with.

**Comment by kiran-chaudhary** [comment]
> It appears that the `-b` and `-e` flags do not work. It always sends null in the request when I try this.

This observation might be because the py42 version might be pointing to `timestamp-formats` branch which didn't support float format then, so this must be fixed now.

**Comment by alanag13** [comment]
> This observation might be because the py42 version might be pointing to timestamp-formats branch which didn't support float format then, so this must be fixed now.

Are you saying that this PR has a dependency on https://github.com/code42/py42/pull/205 ?
This PR already also has a dependency on audit log being available in py42 (it is in master but not merged).

I'm going to convert this PR to draft so that we don't risk merging this before the necessary changes in py42 have been released.

**Comment by kiran-chaudhary** [comment]
> > This observation might be because the py42 version might be pointing to timestamp-formats branch which didn't support float format then, so this must be fixed now.
> 
> Are you saying that this PR has a dependency on [code42/py42#205](https://github.com/code42/py42/pull/205) ?
> This PR already also has a dependency on audit log being available in py42 (it is in master but not merged).
> 
> I'm going to convert this PR to draft so that we don't risk merging this before the necessary changes in py42 have been released.

No @alanag13 , I intended to imply that  @unparalleled-js might have checked against some other Py42 branch which didn't support the timestamp format, as I was aware she had just reviewed my other branch.  Since I was not able to reproduce it, I had a discussion with her on the same yesterday and she confirmed it was some error on her side. 

As per my understanding, once review is completed we are only waiting for Py42 version to be bumped so that we can make appropriate change here.

**Comment by unparalleled-js** [comment]
When there are no results, can we print `no results.` like we do for file events?

**Comment by unparalleled-js** [comment]
The field `requestJson` show up in the output and it really bogs it down and feels unnecessary... Not sure what's it for exactly, but I vote we remove it... It is just a JSON string, so it's like JSON within an ascii Table, kind of weird and kind of like nested fields.

**Comment by unparalleled-js** [comment]
> The field `requestJson` show up in the output and it really bogs it down and feels unnecessary... Not sure what's it for exactly, but I vote we remove it... It is just a JSON string, so it's like JSON within an ascii Table, kind of weird and kind of like nested fields.

Maybe we don't need to remove it.... Not a very tight looking table though. Curious what others think.

**Comment by timabrmsn** [comment]
The `requestJson` gives you the actual FFS query the user made for the search event, which is generally going to be an important part of the auditing. It might be possible to pull out a human-readable form of the filter clauses when showing the table view? Like converting this:

```
{"filters":[{"term":"osHostName","operator":"IS","value":"DESKTOP-H88BEKO","display":null},{"term":"osHostName","operator":"IS","value":"W10E-X64-FALLCR","display":null}],"filterClause":"OR","display":null}
```
to:
```
osHostName is ('DESKTOP-H88BEKO' OR 'W10E-X64-FALLCR')
```

Although that's probably too much effort for our given workload right now, and would likely fit better into py42.

**Comment by alanag13** [comment]
my vote would be to have the table format be the same as it is in the ui and only show the `requestJson` in the csv.
The reason for this is that while audit log _currently_ only is used for ffs searches / downloads, many more types of events that will probably _not_ have a relevant `requestJson` property are planned to be added. The fields displayed in the UI currently are those most likely to be relevant for all current and future event types.

We could consider adding a `detail` option (like what we have for alerts) that would output only the extra fields relevant to the specific type of event.

**Comment by kiran-chaudhary** [comment]
Added message when no results are found, Added default header items for table format.

**Comment by unparalleled-js** [comment]
Looks like this can be ready for review now. Py42 released

**Comment by kiran-chaudhary** [comment]
On a day off, this is the most I could do, just toggled the version and made it available for review :)

**Comment by alanag13** [comment]
@unparalleled-js @kiran-chaudhary the version of py42 that was released yesterday does not contain the audit log functionality -- v1.8.2 was a branch off of v1.8.1 that _only_ added a fix for a paging issue that was in py42.

It does look like the other items have completed testing though, so we should be able to release py42 soon. We will need to bump the py42 version of this again once that happens. I'm going to mark this as draft again until it does to ensure that we don't release too early.

File: setup.py
Comment (unparalleled-js) [inline setup.py:39]: 1.9

**Comment by unparalleled-js** [comment]
Looks good to me, just bump py42 vers

File: integration/test_alerts.py
Comment (kiran-chaudhary) [inline integration/test_alerts.py:1]: Had to fix this, as it used a static date which fails when current date has difference of more than 90 days and removed old sub-commands like `write-to` and `print`. Since the response is dynamic, altered validation to just check the success status of command execution, to make sure at-least the version dependency is verified by real command execution.

File: src/code42cli/cmds/auditlogs.py
Comment (timabrmsn) [inline src/code42cli/cmds/auditlogs.py:182]: I don't really see the benefit of pulling this out into a separate function (`create_simple_send_to_handler`), at least not yet since there's only one place this is used. And there's no need to load the response.text to json as py42 already has already done it in the `Py42Response` object. Unless there's something I'm missing `_send_to()` could be simplified to: 

```python
def _send_to(sdk, hostname, protocol, format, **filter_args):
    logger = get_logger_for_server(hostname, protocol, format)
    with warn_interrupt():
        response_gen = sdk.auditlogs.get_all(**filter_args)
        for response in response_gen:
            if EVENT_KEY in response:
                for event in response[EVENT_KEY]:
                    logger.info(event)
            else:
                click.echo("No results found.")
```

**Review by timabrmsn — CHANGES_REQUESTED** [review]

File: src/code42cli/cmds/auditlogs.py
Comment (kiran-chaudhary) [reply src/code42cli/cmds/auditlogs.py:168]: this `try..except` block is needed for the generator ( `for response in response_gen` ) which does a dict lookup after the `yield` statement in `get_all_pages` method.

File: src/code42cli/cmds/auditlogs.py
Comment (kiran-chaudhary) [reply src/code42cli/cmds/auditlogs.py:90]: added.

File: src/code42cli/cmds/auditlogs.py
Comment (kiran-chaudhary) [reply src/code42cli/cmds/auditlogs.py]: I intended to be consistent here with all "extraction/send-to-server" functions in a single module. Refactored, as requested. Are we fine here with `warn_interrupt` not having a message?

File: src/code42cli/options.py
Comment (kiran-chaudhary) [reply src/code42cli/options.py]: Changed.

File: src/code42cli/cmds/auditlogs.py
Comment (timabrmsn) [reply src/code42cli/cmds/auditlogs.py:167]: Ah, I see. I didn't realize the exception would be raised from py42.. Maybe we should add a check in py42 for the missing key and if it's not there insert an empty list to make it consistent with other endpoints? 

We should also file a bug with the audit logs team about that, I'll look into that part.

File: src/code42cli/cmds/auditlogs.py
Comment (timabrmsn) [reply src/code42cli/cmds/auditlogs.py]: Yeah, but this function doesn't actually use the extractor, so it doesn't quite fit there. 

I'd just rather not try to force abstracting out a function like this until we're sure we'll need that exact functionality in multiple places. 

And warn_interrupt does have a default message (`Cancelling operation cleanly, one moment...`) if none is passed in.

File: src/code42cli/cmds/auditlogs.py
Comment (timabrmsn) [inline src/code42cli/cmds/auditlogs.py:139]: Can we say `to _the_ given server address` to be consistent with the security-data send-to command? And I also noticed when checking to see what the help text was on security-data that `saved-search` is missing help text:

```
Commands:
  clear-checkpoint  Remove the saved file event checkpoint from `--use-checkpoint/-c` mode.
  search            Search for file events.
  saved-search
  send-to           Send events to the given server address.
```

Could you update that too?

Comment (kiran-chaudhary) [reply src/code42cli/cmds/auditlogs.py]: Changed.

File: src/code42cli/cmds/auditlogs.py
Comment (timabrmsn) [inline src/code42cli/cmds/auditlogs.py:28]: Can we add affected username to the table output list? I think anything that we can filter for in the cmd options should be visible in the table, otherwise you can get stuff like this: 

```
# code42 audit-logs search --affected-username "tim.abramson+support@code42.com" -b 90d
Timestamp                  Type   ActorName   ActorIpAddress   Success   ResultCount   
2020-11-19T03:16:18.239Z   None   None        None             None      None          

# code42 audit-logs search --affected-username "tim.abramson+support@code42.com" -b 90d -f JSON
{
    "type$": "audit_log::local_auth_only_changed/1",
    "timestamp": "2020-11-19T03:16:18.239Z",
    "userId": "942564422882759874",
    "userName": "tim.abramson+support@code42.com",
    "newValue": false
}
```

And I see that there are two "type" fields in the results `type$` and `type`, but we're only showing one in the table, but pretty much all events that are not of type 'query' don't have the `type` field. And from the looks of it, Console is using the `type$` field to pull out the event type to display, so I wonder if all new events will only have `type$` going forward, and if we should just switch to displaying that instead of `type`.

Comment (timabrmsn) [reply src/code42cli/cmds/auditlogs.py:28]: Yeah, it also looks like when we use `--event-type` to filter, it only works on the `type$` values, so we should definitely switch to displaying that.

File: src/code42cli/cmds/auditlogs.py
Comment (timabrmsn) [reply src/code42cli/cmds/auditlogs.py:167]: Looks like the bug has already been filed for the empty events result: PL-93211

So I think it's ok to just catch the exception here in the CLI and once the backend is fixed we can revisit if needed.

File: src/code42cli/cmds/auditlogs.py
Comment (kiran-chaudhary) [reply src/code42cli/cmds/auditlogs.py]: Changed, since we can't accommodate all columns in table format, removed `resultCount` and `success` columns and added the requested one.

**Review by timabrmsn — APPROVED** [review]
LGTM

**Pull request merged**
